### PR TITLE
s3 - fix tagging issues

### LINKED ISSUE #33: Ensure idempotency for tags
##### SUMMARY
Make sure idempotency is held for updating tags

##### ISSUE TYPE
- Feature Idea

### PR DESCRIPTION
**Pull request by jatorcasso**
Depends-On: https://github.com/ansible-collections/amazon_cloud_code_generator/pull/37
Depends-On: https://github.com/ansible-collections/amazon_cloud_code_generator/pull/39

##### SUMMARY
Addresses:
- bug converting existing tags
- comparing existing and new tags when lists are not in same order

Fixes #33 

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
s3

### CODE REVIEW

**Comment by alinabuzachis** [comment]
@jatorcasso It should work even without the wait parameter because the next task (the one running idempotency) is supposed to wait the previous one in either way (for this reason, it throws the warning "[WARNING]: There is one or more IN PROGRESS operations on 4424425b-225f-552a-ba49-1607283e2d82. Wait until there are no more IN PROGRESS
operations before proceeding.").
Could you please instead add some other tasks checking that tags work well in the case of idempotency? Thanks.

File: amazon_cloud_code_generator/data/plugins/module_utils/core.py [hunk 1]
```diff
@@ -469,10 +476,10 @@ def update_resource(
                 ) as e:
                     self.module.fail_json_aws(e, msg="Failed to update resource")
 
-                if self.module.params.get("wait"):
-                    self.wait_until_resource_request_success(
-                        response["ProgressEvent"]["RequestToken"]
-                    )
+                # Always wait to update to ensure correct data returned to user
+                self.wait_until_resource_request_success(
+                    response["ProgressEvent"]["RequestToken"]
```
Comment (alinabuzachis) [inline amazon_cloud_code_generator/data/plugins/module_utils/core.py:481]: We should always wait for the resource to be updated.

File: amazon_cloud_code_generator/data/tests/integration/targets/s3/tasks/main.yml [hunk 2]
```diff
@@ -19,119 +19,114 @@
...
 
-    - name: Update S3 bucket public access block configuration and tags - check_mode
+    - name: Update S3 bucket public access block configuration (block_public_policy=true) - check_mode
       amazon.cloud.s3_bucket:
         bucket_name: "{{ output.result.identifier }}"
         state: present
         public_access_block_configuration:
           block_public_acls: false
-          block_public_policy: false
+          block_public_policy: true
```
Comment (alinabuzachis) [inline amazon_cloud_code_generator/data/tests/integration/targets/s3/tasks/main.yml:85]: Why did you change it to true?

Comment (jatorcasso) [reply amazon_cloud_code_generator/data/tests/integration/targets/s3/tasks/main.yml:85]: I think that was a result of me moving the tagging tasks to the `tagging.yml`, i didnt actually modify any of the other tasks

File: amazon_cloud_code_generator/data/plugins/module_utils/core.py (hunk 1 above)
Comment (jatorcasso) [reply amazon_cloud_code_generator/data/plugins/module_utils/core.py:481]: @alinabuzachis if we dont wait, then the data returned to the user can often be incorrect still the update is still in progress. We can choose to not wait, but if we want to return accurate data, we would need to track a `correct` resource to return, rather than using the `get_resource` function without waiting

File: amazon_cloud_code_generator/data/plugins/module_utils/core.py [hunk 3]
```diff
@@ -467,10 +477,10 @@ def update_resource(
                 ) as e:
                     self.module.fail_json_aws(e, msg="Failed to update resource")
 
-                if self.module.params.get("wait"):
-                    self.wait_until_resource_request_success(
-                        response["ProgressEvent"]["RequestToken"]
-                    )
+                # Always wait to update to ensure correct data returned to user
```
Comment (alinabuzachis) [inline amazon_cloud_code_generator/data/plugins/module_utils/core.py:480]: We cannot always wait for updates. It there's an issue with tag update, we should handle singularly!

**Comment by alinabuzachis** [comment]
@jatorcasso It seems it also fails when purge_tags is false. It's not additive.

```
TASK [s3 : test updating tags without purge on amazon.cloud.s3_bucket] **********************************************************
task path: /Users/alinabuzachis/dev/collections/ansible_collections/amazon/cloud/tests/output/.tmp/integration/s3-trgqwidi-ÅÑŚÌβŁÈ/tests/integration/targets/s3/tasks/tagging.yml:145
Using module file /Users/alinabuzachis/dev/collections/ansible_collections/amazon/cloud/plugins/modules/s3_bucket.py
Pipelining is enabled.
<testhost> ESTABLISH LOCAL CONNECTION FOR USER: alinabuzachis
<testhost> EXEC /bin/sh -c 'ANSIBLE_DEBUG_BOTOCORE_LOGS=True /Users/alinabuzachis/anaconda3/envs/py38/bin/python && sleep 0'
changed: [testhost] => {
    "changed": true,
    "invocation": {
        "module_args": {
            "accelerate_configuration": null,
            "access_control": null,
            "analytics_configurations": null,
            "aws_access_key": "xxxx",
            "aws_ca_bundle": null,
            "aws_config": null,
            "aws_secret_key": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "bucket_encryption": null,
            "bucket_name": "d1e0f3d0-648e-57a8-a211-df6748b02253",
            "cors_configuration": null,
            "debug_botocore_endpoint_logs": true,
            "ec2_url": null,
            "force": false,
            "intelligent_tiering_configurations": null,
            "inventory_configurations": null,
            "lifecycle_configuration": null,
            "logging_configuration": null,
            "metrics_configurations": null,
            "notification_configuration": null,
            "object_lock_configuration": null,
            "object_lock_enabled": null,
            "ownership_controls": null,
            "profile": null,
            "public_access_block_configuration": null,
            "purge_tags": false,
            "region": "us-east-1",
            "replication_configuration": null,
            "security_token": null,
            "state": "present",
            "tags": {
                "CamelCaseKey": "CamelCaseValue",
                "Key with Spaces": "Value with spaces",
                "New Key with Spaces": "Updated Value with spaces",
                "pascalCaseKey": "pascalCaseValue",
                "snake_case_key": "snake_case_value"
            },
            "validate_certs": true,
            "versioning_configuration": null,
            "wait": false,
            "wait_timeout": 320,
            "website_configuration": null
        }
    },
    "resource_actions": [
        "cloudcontrolapi:UpdateResource",
        "cloudcontrolapi:ListResourceRequests",
        "cloudcontrolapi:GetResource"
    ],
    "result": {
        "identifier": "d1e0f3d0-648e-57a8-a211-df6748b02253",
        "properties": {
            ...,
            "tags": {
                "New Key with Spaces": "Value with spaces",
                "NewCamelCaseKey": "CamelCaseValue",
                "newPascalCaseKey": "pascalCaseValue",
                "new_snake_case_key": "snake_case_value"
            },
            ...
        }
    }
}

TASK [s3 : assert that update succeeded] ****************************************************************************************
task path: /Users/alinabuzachis/dev/collections/ansible_collections/amazon/cloud/tests/output/.tmp/integration/s3-trgqwidi-ÅÑŚÌβŁÈ/tests/integration/targets/s3/tasks/tagging.yml:150
fatal: [testhost]: FAILED! => {
    "assertion": "_result.result.properties.tags == final_tags",
    "changed": false,
    "evaluated_to": false,
    "msg": "Assertion failed"
}


```

**Comment by alinabuzachis** [comment]
Superseeded by https://github.com/ansible-collections/amazon_cloud_code_generator/pull/45
